crates/core/src/delta_datafusion/cdf/mod.rs

```diff
@@ -0,0 +1,79 @@
+//! Logical operators and physical executions for CDF
+
+use std::collections::HashMap;
+
+pub(crate) use scan::*;
+pub(crate) use scan_utils::*;
+
+use crate::kernel::{Add, AddCDCFile, Remove};
+
+mod scan;
+mod scan_utils;
+
+#[derive(Debug)]
+pub(crate) struct CdcDataSpec<F: FileAction> {
+    version: i64,
+    timestamp: i64,
+    actions: Vec<F>,
+}
+
+impl<F: FileAction> CdcDataSpec<F> {
+    pub fn new(version: i64, timestamp: i64, actions: Vec<F>) -> Self {
+        Self {
+            version,
+            timestamp,
+            actions,
+        }
+    }
+}
+
+/// This trait defines a generic set of operations used by CDF Reader
+pub trait FileAction {
+    /// Adds partition values
+    fn partition_values(&self) -> HashMap<String, Option<String>>;
+    /// Physical Path to the data
+    fn path(&self) -> String;
+    /// Byte size of the physical file
+    fn size(&self) -> usize;
+}
+
+impl FileAction for Add {
+    fn partition_values(&self) -> HashMap<String, Option<String>> {
+        self.partition_values.clone()
```

Comment:
seems like you only use this to `get` out of the map, so you could probably return `&HashMap<String, Option<String>>` and not have to clone (or just provide a `get_partition_value` which does the probe internally (which makes it easier to not have to clone for remove).